architecture-decision-records/0022-cloudfront-upload-proxy.md

```diff
@@ -0,0 +1,164 @@
+the file and sending the start and end requests to combine the chunks in S3.
+
+We considered running NGINX in ECS, or writing proxy code and running it in Lambda or ECS. We didn't look very far into
+these options because it means maintaining an extra service. We would also have to proxy the requests to Cognito and
+STS, or find some way to use the AWS SDK to send Keycloak tokens with the upload requests.
+
+We also looked into [S3 website endpoints], which allow you to assign a custom domain in Route 53 to an S3 bucket. This
+was definitely not an option because it only supports requests over plain HTTP. You cannot make requests using HTTPS
+URLs because there is nowhere to add a certificate for SSL termination.
+
+The main option we looked at was using CloudFront as an upload proxy. We built a spike in the Sandbox environment to
+check that it would work (see [TDR-1261]). The main reason that this didn't stand out as the obvious option from the
+start is that most documentation around CloudFront and S3 assumes that you want to use them to host a static website, so
+it was tricky to find the right guides to using CloudFront as an upload proxy. But this does seem to be an expected way
+to use CloudFront, so it looks like a good option for us.
+
+### Authorisation options
+
+We considered these options for restricting users' permissions to upload files:
+
+- Keep the current system of Keycloak, Cognito, and STS tokens
+- Use the user's Keycloak access token directly
+- Use one of CloudFront's systems for restricting content: signed cookies or signed URLs
+
+Keeping the existing token exchange process would have meant proxying two extra requests: one to Cognito and one to STS.
+It's already quite awkward to have to use three sets of tokens (Keycloak, Cognito and STS).
+
+Another disadvantage of using Cognito tokens for S3 authorisation is that the IAM policy has to use the Cognito ID as
+part of the object key to restrict S3 uploads to the user's own S3 "folder". This would be fine, except we have no easy
+way to link a Keycloak user ID to a Cognito ID. Cognito will only generate a user ID when given a valid token for the
+Keycloak user, which requires the user to authenticate themselves, so an admin user cannot convert an arbitrary Keycloak
+ID to a Cognito ID or vice-versa.
+
+Using just the Keycloak token would simplify the workflow. If we use CloudFront, we could configure an [edge function]
+to validate the token and authorise the request. The downside is that it's not obviously compatible with the AWS
+JavaScript SDK. The SDK sets an `Authorization` header based on AWS credentials. It's not clear how you can add a custom
+header based on the Keycloak token. So if we wanted to use just the Keycloak tokens to authorise the S3 upload, we would
+have to write more client-side JavaScript to handle multipart uploads to replace the call to the SDK. It would still be
+worth considering this option if the assumptions change in future, though.
```

Comment:
Maybe delete the "..., though" from the end of the sentence. Not sure its needed.

Reply:
👍